source/opt/trim_capabilities_pass.cpp

```diff
@@ -0,0 +1,313 @@
+}};
+
+// ==============  End opcode handler implementations.  =======================
+
+namespace {
+ExtensionSet getExtensionsRelatedTo(const CapabilitySet& capabilities,
+                                    const AssemblyGrammar& grammar) {
+  ExtensionSet output;
+  const spv_operand_desc_t* desc = nullptr;
+  for (auto capability : capabilities) {
+    if (SPV_SUCCESS != grammar.lookupOperand(SPV_OPERAND_TYPE_CAPABILITY,
+                                             static_cast<uint32_t>(capability),
+                                             &desc)) {
+      continue;
+    }
+
+    for (uint32_t i = 0; i < desc->numExtensions; ++i) {
+      output.insert(desc->extensions[i]);
+    }
+  }
+
+  return output;
+}
+}  // namespace
+
+TrimCapabilitiesPass::TrimCapabilitiesPass()
+    : supportedCapabilities_(
+          TrimCapabilitiesPass::kSupportedCapabilities.cbegin(),
+          TrimCapabilitiesPass::kSupportedCapabilities.cend()),
+      forbiddenCapabilities_(
+          TrimCapabilitiesPass::kForbiddenCapabilities.cbegin(),
+          TrimCapabilitiesPass::kForbiddenCapabilities.cend()),
+      untouchableCapabilities_(
+          TrimCapabilitiesPass::kUntouchableCapabilities.cbegin(),
+          TrimCapabilitiesPass::kUntouchableCapabilities.cend()),
+      opcodeHandlers_(kOpcodeHandlers.cbegin(), kOpcodeHandlers.cend()) {}
+
+void TrimCapabilitiesPass::addInstructionRequirements(
+    Instruction* instruction, CapabilitySet* capabilities,
+    ExtensionSet* extensions) const {
+  // First case: the opcode is itself gated by a capability.
+  {
+    const spv_opcode_desc_t* desc = {};
+    auto result =
+        context()->grammar().lookupOpcode(instruction->opcode(), &desc);
+    if (result == SPV_SUCCESS) {
+      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities,
+                                    capabilities);
+    }
+
+    if (desc->minVersion <= spvVersionForTargetEnv(context()->GetTargetEnv())) {
+      extensions->insert(desc->extensions,
+                         desc->extensions + desc->numExtensions);
+    }
+  }
+
+  // Second case: one of the opcode operand is gated by a capability.
+  {
+    const uint32_t operandCount = instruction->NumOperands();
+    for (uint32_t i = 0; i < operandCount; i++) {
+      const auto& operand = instruction->GetOperand(i);
+      // No supported capability relies on a 2+-word operand.
+      if (operand.words.size() != 1) {
+        continue;
+      }
+
+      // No supported capability relies on a literal string operand.
+      if (operand.type == SPV_OPERAND_TYPE_LITERAL_STRING) {
+        continue;
+      }
+
+      const spv_operand_desc_t* desc = {};
+      auto result = context()->grammar().lookupOperand(operand.type,
+                                                       operand.words[0], &desc);
+      if (result != SPV_SUCCESS) {
+        continue;
+      }
+
+      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities,
```

Comment:
This is the new location of my earlier comment.

In the first case, you have the code:

```
      extensions->insert(desc->extensions,
                         desc->extensions + desc->numExtensions);
```

but that is not replicated here. I there a reason for that?

Reply:
There is, AFAIK only one extension enabled by an operand: VulkanMemoryModel. We don't support it for now, so this path is not tested.
To stay coherent, I added it to this loop, but keeping the handler capability-only until we have test cases (so a supported capability requiring this).